[JENKINS-35095] Findbugs issues. Migrate to 2.9 parent pom.

armfergom opened the pull request:
- Migrate to 2.9 parent pom
- Findbugs issues
- Injected test fix

@reviewbybees

jglick commented on src/main/java/org/jenkinsci/plugins/credentialsbinding/Binding.java:
```diff
@@ -77,7 +79,7 @@ public SingleEnvironment(String value, Unbinder unbinder) {
 
     @Deprecated
-    public interface Environment {
+    public interface Environment extends Serializable {
```
Why?

armfergom replied:
`UnbinderWrapper` has a [field](https://github.com/jenkinsci/credentials-binding-plugin/blob/master/src/main/java/org/jenkinsci/plugins/credentialsbinding/Binding.java#L115) of type `Environment`. `UnbinderWrapper` implements `Unbinder` and `Unbinder` implements `Serializable`. `Environment` interface does not implement `Serializable` and Findbugs complains about that with the following error:

> [INFO] Class org.jenkinsci.plugins.credentialsbinding.Binding$UnbinderWrapper defines non-transient non-serializable instance field e [org.jenkinsci.plugins.credentialsbinding.Binding$UnbinderWrapper] In Binding.java SE_BAD_FIELD

jglick replied:
You just need to suppress that warning. `Environment` is deprecated for a reason: it was not, in general, serializable. `UnbinderWrapper` is used only if a `Binding` subtype failed to be updated to support Pipeline, still implementing `bind`, and such implementations would typically have an unserializable environment.

jglick commented on src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java:
```diff
@@ -146,9 +148,9 @@
             return new LineTransformationOutputStream() {
                 @Override protected void eol(byte[] b, int len) throws IOException {
-                    Matcher m = p.matcher(new String(b, 0, len));
+                    Matcher m = p.matcher(new String(b, 0, len, Charsets.UTF_8));
```
:bug: this should probably be using `Run.getCharset`. (As noted in https://github.com/jenkinsci/jenkins/pull/2302, it is hard to test these things realistically.)

armfergom commented:
@jglick addressed your comments. 

- `Environment` no longer extends `Serializable` and Findbugs error is ignored.
- Charset is retrieved from the `Run`. Since the `AbstractBuild`in `decorateLogger` is `null` (as stated [here](https://github.com/jenkinsci/workflow-step-api-plugin/blob/master/src/main/java/org/jenkinsci/plugins/workflow/steps/BodyInvoker.java#L109)), the charset name needs to be passed as parameter in `Filter` constructor. Nor the `Run` or the `Charset` can be a field in `Filter` class since they are not serializable, causing problems when executing the step. Please, advise if you had something different in mind.

jglick commented on src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java:
```diff
@@ -140,15 +143,16 @@
             return new LineTransformationOutputStream() {
                 @Override protected void eol(byte[] b, int len) throws IOException {
-                    Matcher m = p.matcher(new String(b, 0, len));
+                    Matcher m = p.matcher(new String(b, 0, len, Charset.forName(charsetName)));
```
:bug: this will throw an `IllegalArgumentException("Null charset name")` when resuming a build that had been running while you restarted Jenkins to update this plugin.

In general, when introducing a field to a `Serializable` class, you must consider the possibility that it resolves to `null`.

armfergom replied:
I have a fix for this. But before pushing it I wanted to test it. I have not been able to reproduce the issue at all with the plugin as it is in this PR. Had a job using `withCredentials` that sleeps for 100 seconds using latest version of Credentials Binding. While the job is running I upgrade the plugin manually to the one in this PR, restart Jenkins instance and all is good. Job finishes successfully but it should not, right? Did I misunderstand something?

armfergom replied:
Actually, I have anyway pushed what I think would fix the issue. When I am able to reproduce it I'd like to write a test for it...

jglick replied:
Note that you could use just `charsetName` rather than `Charset.forName(charsetName)`.

jglick commented on pom.xml:
```diff
@@ -17,16 +16,20 @@
+    <jenkins.version>1.596.1</jenkins.version>
+    <java.level>6</java.level>
   </properties>
+  
```
Please avoid trailing whitespace.

jglick commented on src/main/java/org/jenkinsci/plugins/credentialsbinding/Binding.java:
```diff
@@ -25,12 +25,16 @@
 import hudson.model.AbstractBuild;
 import hudson.model.BuildListener;
 import java.io.IOException;
+import java.io.Serializable;
```
:ant: unused import

jglick commented:
:bee:

armfergom commented:
Addressed your comments (unused imports, trailing whitespaces, and other suggestions) and squashed commits.

As for the potential issue with `charsetName` field, I agree we must cover that corner case, but it seems to me that `decorateLogger` (where `charsetName` is used) is only called when the [job is started](https://github.com/jenkinsci/credentials-binding-plugin/blob/master/src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java#L95) but not ever again (or if it is resumed after a restart). I might be wrong, but I think that is why I was not able to reproduce the issue.

Thanks for the review.

jglick commented:
> and squashed commits

I would prefer you _not_ make destructive history operations in a pull request. Among other things, it makes it impossible to see what you changed since my last review.

jglick commented:
`decorateLogger` is called when the _step_ is started. The _build_ may have started days earlier, with several Jenkins restarts intervening.

Anyway the issue I hinted at would arise when you start the step with the old plugin version, then restart Jenkins with the new plugin version.

jglick commented:
re:bee:
